packages/fluentui/react-bindings/src/hooks/useClasses.ts

```diff
@@ -0,0 +1,218 @@
+  /** A classname that will be added by default to all instances of a component on the `root` slot. */
+  className?: string;
+
+  /**
+   * A mapping from component's props to styles functions props. Can be only primitive types as they will be used for
+   * cache keys.
+   */
+  mapPropsToStyles?: () => StyleProps;
+
+  /**
+   * A set props of that contain mapping for props that perform inline styles overrides, for example `styles` or
+   * `variables`.
+   */
+  mapPropsToInlineStyles?: () => InlineStyleProps<StyleProps>;
+};
+
+type InlineStyleProps<StyleProps> = {
+  /** Additional CSS class name(s) to apply.  */
+  className?: string;
+
+  design?: ComponentDesignProp;
+
+  /** Additional CSS styles to apply to the component instance.  */
+  styles?: ComponentSlotStyle<StyleProps, any>; // TODO: see if we can improve it
+
+  /** Override for theme site variables to allow modifications of component styling via themes. */
+  variables?: ComponentVariablesInput;
+};
+
+export const useClasses = <StyleProps extends PrimitiveProps>(
+  displayName: string,
+  options: UseStylesOptions<StyleProps>,
+): string => {
+  const { disableAnimations, performance: performanceFlags, telemetry, theme, renderer, rtl } = useFluentContext();
+  const {
+    className: componentClassName = process.env.NODE_ENV === 'production' ? '' : 'no-classname-🙉',
+    mapPropsToStyles = () => ({} as StyleProps),
+    mapPropsToInlineStyles = () => ({} as InlineStyleProps<StyleProps>),
+  } = options;
+
+  const componentStylesProps = mapPropsToStyles();
+  const inlineStylesProps = mapPropsToInlineStyles();
+
+  //
+  // To compute styles we are going through three stages:
+  // - resolve variables (siteVariables => componentVariables + props.variables)
+  // - resolve styles (with resolvedVariables & props.styles & props.design)
+  // - compute classes (with resolvedStyles)
+  // - conditionally add sources for evaluating debug information to component
+
+  const telemetryVariablesStart = telemetry?.enabled ? performance.now() : 0;
+
+  let resolvedVariablesHash: string = '';
+  let resolvedVariables: ComponentVariablesObject = {};
+
+  if (typeof theme.componentVariables[displayName] === 'function') {
+    // @ts-ignore
+    if (theme.componentVariables[displayName].__hash) {
+      // @ts-ignore
+      resolvedVariables = theme.componentVariables[displayName].__resolved;
+      // @ts-ignore
+      resolvedVariablesHash = theme.componentVariables[displayName].__hash;
+    } else {
+      resolvedVariables = theme.componentVariables[displayName](theme.siteVariables);
+      resolvedVariablesHash = hash(JSON.stringify(resolvedVariables));
+
+      // @ts-ignore
+      theme.componentVariables[displayName].__resolved = resolvedVariables;
+      // @ts-ignore
+      theme.componentVariables[displayName].__hash = resolvedVariablesHash;
+    }
+  } else if (typeof theme.componentVariables[displayName] !== 'undefined') {
+    throw new Error();
+  }
```

Comment:
Maybe I'm remembering this wrong... but can't componentVariables be an object too ? 

Also there is no message in this error, it might be really hard to understand what is going when this throws if you are consuming it